### DISCUSSION TITLE
Parsing Slingbox .bin files, for "Tivo Unit Addresses"

### DISCUSSION THREAD

**Discussion by RandomMan719** [GerryDazoo/Slinger#262] [Q&A]
Does anyone know how to parse a Slingbox remote .bin file, so we can see the keycodes?

**Comment by DVoorhees1957** [comment]
Just posted this on the other thread, but thought I'd copy it here in case it triggers something for someone:

It's gotta be something simple, the BIN files for the S3 (unit address 1) and S3 (unit address 2) only differ in two places: offset 3 (x'DC' vs x'9D') and offset 15 (x'8B' vs x'9B')... it suggests the 1st 16 bytes of the BIN file are "preamble" which is used to determine which TiVo's will receive and act on the commands, and the last 37 bytes are keycode mappings (which are basically identical for all TiVos, ignoring any "TiVo unit address")...

**Reply by RandomMan719** [reply]
I am going to add Tivo Unit Address to this thread, yeah I don't understand how it works.  It has to send different codes for each unit address, I assume.

**Reply by RandomMan719** [reply]
This might be helpful, maybe it is something that has to be done in the slingbox_server.py script.  IDK

http://www.hifi-remote.com/forums/dload.php?action=file&file_id=4805

**Reply by RandomMan719** [reply]
This text is from the TiVo-S3-1718.rmdu file

"Notes=This upgrade allows you to more effectively use multi-unit TiVo support with the S3 and Sling.\n\nBy default Sling only includes multi-unit support for TiVo S2 (1-4).  While you can use the S2 multi-unit support for the S3, the "Clear" and "Aspect"(stdby) buttons do not work correctly.\n\nThis upgrade fixes those buttons for multi-unit TiVo S3 support and also allows you to specify any unit \#s TiVo supports.  I have not found a way to associate the TiVo S3 graphical remote image with this upgrade, so the image of the remote will still be the old S2 remote.\n\nThis upgrade allows you to customize the Unit IDs to anything supported by TiVo.  By default Sling only includes TiVo devices for unit 0(all units), 1, 2, 3, 4.\n\nSet the "Remote" to Slingbox with UEI PL or JU\nSet the "Unit Code"\nChoose "Advanced"->"Export Binary Upgrade"\nPlace files in your SlingPlayer\\SBAV directory\nDefault filename is V1718_PL.bin\n\nYou can select to use your new .bin file by using the Slingbox client\n\nSlingbox->Properties->Slingbox Configuration->Audio/Video Setup\n->DVR (Digital Video Recorder)->TiVo->Other->Alternative Code\n->V1718\n"

**Comment by RandomMan719** [comment]
@DVoorhees1957 Did you ever get this figured out?

**Reply by DVoorhees1957** [reply]
No, I'm now using a different iPhone-based TiVo solution that communicates directly with my TiVo to send remote codes, rather than going through Slinger, so I haven't spent any energy on this. Plus, with all of the chaos that Sling created by bricking some of the boxes, I didn't want to take the focus off of finding solutions for un-bricking the boxes.

The BIN files for remotes are so incredibly small (50-60 bytes) that there has to be some fixed format or structure to them. But I have no insight into that. It might help to enlist the assistance of someone who has actually fiddled with those BIN files, whether for TiVos or some other devices.

A better long-term solution, of course, is to develop a Windows/Mac app that can successfully reconfigure the Slingbox input (and the associated remote) without any dependency on communicating with Sling's now-defunct servers. That would remove any need to parse/interpret the remote BIN files themselves...

**Reply by RandomMan719** [reply]
SlingPlayer 1.0.10 on Macintosh can scan digital channels, configure all inputs, and add custom remotes, all Offline with no communication with Mr. Sling's Servers.

**Reply by DVoorhees1957** [reply]
Cool... do we have a corresponding Windows version of Slingplayer that can do the same thing? No need to reinvent the wheel if we don't have to...

**Reply by RandomMan719** [reply]
I do want to add that I have only been testing the SlingPlayer 1.0.10 Mac, using a ProHD.  I am not sure if the newer models work with that version, but I can test that in a few days.  I think it is limited to the Pro/ProHD/Solo.  I haven't been able to do a digital scan on PC using SlingPlayer 1.5.x. which works without login into the servers.

**Reply by DVoorhees1957** [reply]
There seem to be some "BIN conversion" tools in the "Slingbox Work Files" directory of the JP1 Remotes site; i.e., at the following URL: http://www.hifi-remote.com/forums/dload.php?action=category&cat_id=77. For instance, see the "SBAV decrypter" (http://www.hifi-remote.com/forums/dload.php?action=file&file_id=3318) and the "Upgrade to BIN converter" (http://www.hifi-remote.com/forums/dload.php?action=file&file_id=2037). However, when I go to those URLs, I don't see any download links so I can't tell if there is anything useful there.

**Comment by DVoorhees1957** [comment]
I finally got back around to this, and found the primary tool used in conjunction with the JP1 site. It's called RemoteMaster, aka RM, RMIR and/or RMDU. Primary reference resources are:

http://www.hifi-remote.com/wiki/index.php/How_to_Create_Custom_Remote_Controls_and_use_them_on_your_Slingbox

and

http://www.hifi-remote.com/wiki/index.php?title=JP1_-_Just_How_Easy_Is_It%3F_-_RM-IR_Version

I've managed to figure out how to load a Slingbox BIN file, and will take some screen shots and add them here as replies to show what's the same and what's different between two TiVo BIN files, the 1st with Unit Address 1 and the 2nd with Unit Address 2.

**Reply by DVoorhees1957** [reply]
Here are the screen shots for V1719_PL.BIN, which is the BIN file for a TiVo Series 3 with unit address 1 (note: tab 3 is blank for these BIN files, so I didn't include a screen shot):

![V1719_PL TiVo S3 unit code 1 - tab 1](https://user-images.githubusercontent.com/13420879/205450906-4ae843cc-c22b-49b4-b7ed-b084d3c1c3d7.jpg)
![V1719_PL TiVo S3 unit code 1 - tab 2](https://user-images.githubusercontent.com/13420879/205450917-18bc1d6b-b07d-4329-bba9-49385a1f3305.jpg)
![V1719_PL TiVo S3 unit code 1 - tab 4](https://user-images.githubusercontent.com/13420879/205450980-451f5407-1229-44c9-98f6-312a79bf3c19.jpg)
![V1719_PL TiVo S3 unit code 1 - tab 5](https://user-images.githubusercontent.com/13420879/205450984-a4a5a98d-d674-453e-9362-d37d2a86819d.jpg)
![V1719_PL TiVo S3 unit code 1 - tab 6](https://user-images.githubusercontent.com/13420879/205450996-fffa078c-9cd7-491f-87a4-a19fa6c848db.jpg)
![V1719_PL TiVo S3 unit code 1 - tab 7](https://user-images.githubusercontent.com/13420879/205451019-ff7a3c42-31a7-447f-944f-b97d22f3ba4d.jpg)
![V1719_PL TiVo S3 unit code 1 - BIN file contents](https://user-images.githubusercontent.com/13420879/205451026-fe3d8094-b5d9-42d6-af28-aa63fdd1b9ee.jpg)

And here are the screen shots for V1720_PL.BIN, which is the BIN file for a TiVo Series 3 with unit address 2 (note: the TiVo unit address, aka "unit code", seems to appear on the 1st tab - I haven't tried to figure out how it changes the keycodes, if at all, or how it gets transmitted):

![V1720_PL TiVo S3 unit code 2 - tab 1](https://user-images.githubusercontent.com/13420879/205451809-9d6ad157-9d78-4145-bee2-47fa59e963b6.jpg)
![V1720_PL TiVo S3 unit code 2 - tab 2](https://user-images.githubusercontent.com/13420879/205451819-b5ca6286-5a16-4d85-a1df-ebc51c4c23ff.jpg)
![V1720_PL TiVo S3 unit code 2 - tab 4](https://user-images.githubusercontent.com/13420879/205451830-6861b96d-8ed4-4c99-b164-6d1940c2ff0a.jpg)
![V1720_PL TiVo S3 unit code 2 - tab 5](https://user-images.githubusercontent.com/13420879/205451844-8a021241-a4a0-4afd-9b79-372ed2e50300.jpg)
![V1720_PL TiVo S3 unit code 2 - tab 6](https://user-images.githubusercontent.com/13420879/205451853-c6072156-409b-47c5-9280-cf3c4e8fb4f2.jpg)
![V1720_PL TiVo S3 unit code 2 - tab 7](https://user-images.githubusercontent.com/13420879/205451863-74d70cf0-279b-484b-b529-e158e3b34d26.jpg)
![V1720_PL TiVo S3 unit code 2 - BIN file contents](https://user-images.githubusercontent.com/13420879/205451873-5c548715-fd8a-4fe2-baab-4dfbeccb8e1d.jpg)
